### ISSUE TITLE
Accepting an offer multiple times has issues

### ISSUE THREAD

**Issue by Christewart** [bitcoin-s/bitcoin-s#3846]
**Describe the bug**

I tried to accept an offer when testing #3839 and it failed to process. Unfortunately our wallet leaves things in a corrupted state when this happens. For instance, inside of [`initDLCAccept()`](https://github.com/Christewart/bitcoin-s-core/blob/d6646463aacf647faec9d4bc1d1108f508dc2988/dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala#L448) you see there are multiple database writes done. 

`dlcId` (the primary key for `global_dlc_data` table) is calculated deterministically based on the funding inputs: https://github.com/Christewart/bitcoin-s-core/blob/d6646463aacf647faec9d4bc1d1108f508dc2988/dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala#L468

This is where we insert the `DLCDb` into the database 

https://github.com/Christewart/bitcoin-s-core/blob/d6646463aacf647faec9d4bc1d1108f508dc2988/dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala#L511

If something fails after this, our database ends up in such a state that those utxos can never be re-used again from the offerer. This is because we will constantly violate the foreign key constraint in `DLCTable`

https://github.com/Christewart/bitcoin-s-core/blob/656e9b1b5de7c93eefe8ae080b05b9cd12d4fc62/dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/models/DLCDAO.scala#L120

This is probably related to #3844 

#### Solution

We need to make all the database writes atomic. I don't think there is a way to get around this.

**Comment by Christewart** [comment]
These utxos are the ones being selected

```
[info] 2021-12-02T14:52:48UTC INFO [DLCWallet$DLCWalletImpl] Creating DLC Offer
[info] 2021-12-02T14:52:50UTC INFO [DLCWallet$DLCWalletImpl] Spending UTXOs: d87fedabac6c804a75afcdc8fccb93315e9854a10487cdf31407adaff98b69c4:1, aac015dd33304ee4d017daa9916870c57ccf8a86fbf93500cee500507f7c138c:91, aac015dd33304ee4d017daa9916870c57ccf8a86fbf93500cee500507f7c138c:105, c0e3238d4c91225c9548bc1c05d710418a7dadc9b72c2bc7ec94834b15f1123a:0, f10f342381fbb68e03c77ad89e44bc5b03c9f1ec6147de2f5b34615639eea152:2, 0e6052401f1f6efcbb7e8ef57d76c4a999c675ec73dc46c0bacf282b107145f4:0, df82ffc4ed4b663ddd2374481e4f16170d51758b9965b5da20464492846c58d9:0, c2535c7d36212c0bf3478584a32494dc0b0a3c1a567edc53838dd470d6d5e6bb:0, 050f4cb35f2cc239f459f74ab6583b89114d3c2411b73fb58354fe72258c1e0d:0
[info] 2021-12-02T14:52:50UTC INFO [DLCWallet$DLCWalletImpl] UTXO 0 details: TransactionOutput(1179 sats,wpkh(52584eb56f12d421e5efc49ecddc5eb873958cc0))
[info] 2021-12-02T14:52:50UTC INFO [DLCWallet$DLCWalletImpl] UTXO 1 details: TransactionOutput(547 sats,wpkh(41aef0c553260b7316412f74086eba43ac7a4caa))
[info] 2021-12-02T14:52:50UTC INFO [DLCWallet$DLCWalletImpl] UTXO 2 details: TransactionOutput(547 sats,wpkh(b209ba581834315d281109b546b6778b26cbd765))
[info] 2021-12-02T14:52:50UTC INFO [DLCWallet$DLCWalletImpl] UTXO 3 details: TransactionOutput(3035 sats,wpkh(032f4c093ad59c5a2ebb7ce447835205f5ecb5b5))
[info] 2021-12-02T14:52:50UTC INFO [DLCWallet$DLCWalletImpl] UTXO 4 details: TransactionOutput(10667 sats,wpkh(e7e2dab5dbc6fa35e60881802ef5e23fc4a92f51))
[info] 2021-12-02T14:52:50UTC INFO [DLCWallet$DLCWalletImpl] UTXO 5 details: TransactionOutput(5953 sats,wpkh(f86785b71f4f44530c009a2f14e8109927733890))
[info] 2021-12-02T14:52:50UTC INFO [DLCWallet$DLCWalletImpl] UTXO 6 details: TransactionOutput(2831 sats,wpkh(107eb2463ad25843ed01fd33656c83bdd11db355))
[info] 2021-12-02T14:52:50UTC INFO [DLCWallet$DLCWalletImpl] UTXO 7 details: TransactionOutput(12358 sats,wpkh(48ba92affa20e46dc7c2fcec746080889ac78492))
[info] 2021-12-02T14:52:50UTC INFO [DLCWallet$DLCWalletImpl] UTXO 8 details: TransactionOutput(14195 sats,wpkh(e60ce491b7f7a548ed6d6b8ca88d47ad398bdfed))
```

**Comment by Christewart** [comment]
Saw this today with a user on 2066447c . This totally gets swallowed and doesn't bubble anything up to the UI layer information a user that a DLC failed. Even worse, it doesn't unreserved utxos causing the users balance to drain over time. 

Maybe some easy pieces to tackle first before doing the huge migration is

1. Unreserve utxos if this fails
2. Bubble the error message to the UI.

and then do the heavy lifting to fix the root cause.

```
2022-02-03T22:25:20UTC ERROR [DLCDataHandler] Failed to process lnMessage=LnMessage(DLCOfferTLV(None,0,DoubleSha256Digest(6fe28c0ab6f1b372c1a6a246ae63f74f931e8365e15a089c68d6190000000000),ContractInfoV0TLV(20000 sats,ContractDescriptorV0
TLV(Vector((Even,20000 sats), (Odd,0 sats))),OracleInfoV0TLV(OracleAnnouncementV0TLV(SchnorrDigitalSignature(SchnorrNonce(ByteVector(32 bytes, 0x511fe31d69dc5c938769958bd5861bf55cbdc180d238696e708f803a32b50219)),FieldElement(ByteVector(3
2 bytes, 0x31ea104f19662348a4699d046383a02204d69f23de614af24bfbb9e95dac8405))),SchnorrPublicKey(ByteVector(32 bytes, 0x1d5dcdba2e64cb116cc0c375a0856298f0058b778f46bfe625ac6576204889e4)),OracleEventV0TLV(IndexedSeq(SchnorrNonce(ByteVector
(32 bytes, 0x317ee600725aeee7dd5fad29742ef6de2e2b17324b3906bc79b0e2f8d87b4aea))),UInt32Impl(1644019200),EnumEventDescriptorV0TLV(Vector(Even, Odd)),Will blockheight 721710 end in an even or odd bit)))),ECPublicKey(03591285d3a984ed2428954
c8a5d35bbbcf6770038f73c55aef52f3fe02c67ef3f),wpkh(787ffa70b17c5004ab41b710da02438096c3149b),UInt64Impl(10041039206265875351),10000 sats,Vector(FundingInputV0TLV(UInt64Impl(14754716440616037340),WitnessTransaction(Int32Impl(2),Vector(Tran
sactionInputImpl(TransactionOutPoint(ee0602286836cffac5b732ca33bba34b090eb9119245c8049eacc0b8a3ba14c2:0),EmptyScriptSignature,UInt32Impl(4294967293)), TransactionInputImpl(TransactionOutPoint(b659419691dcc744b19aeec14eac7fb2a32327f41c5f8
be31e96ef1fb23528c6:1),EmptyScriptSignature,UInt32Impl(4294967293)), TransactionInputImpl(TransactionOutPoint(b4790e849be1e97a59ff0a9b5245bff4dc8b82afc0275d8713e6fbf6cb40a099:0),EmptyScriptSignature,UInt32Impl(4294967293))),Vector(Transa
ctionOutput(482204 sats,wpkh(7ac6e7689c8019821e92bde8ef8a5d4585d5e9a6)), TransactionOutput(75170 sats,wsh(f4734eaadf8ab4c6d3b4c7bc3aa6a66d1112982e2b97f596c0c09428d3f2a22a)), TransactionOutput(19721 sats,wpkh(ac9eca0fb2c9b9485e00807ad1b88
c46d106c875))),UInt32Impl(0),IndexedSeq(P2WPKHWitnessV0(List(027dab791253075cc4894cfa528ef4fde9e9e3553d65b2c23055f3d5472943ef59, 304402202f8d8b93f6878d3ed472b1d1aca51d1bd68751630939bb2a8a261e8296398ba902201e8c029f57972bdcb6f2b60108e70845
98d23f016dbeb39c3867c05cb474472701)), P2WPKHWitnessV0(List(03ad01400bf28765354dae625732d53d10067a2ddf88f3e2443260462e6646d3b3, 304402200b5daf12dcfc6c9c28527da764122aef7a1860d9a085e3162c9c858d14eccb8c02203122adeebf196c205582c455a9bc330a97
4a978ad71b3ecaec40b9bebc32375701)), P2WPKHWitnessV0(List(02044461e3023c3153606404c193de2b7c555297c8faa364b339695e5b35c29242, 3044022001425b09400fb0bf04256f4985ef9499abe125cc5f472231c4c6d8acd8dd2df9022007688da37c62174bc3b470cd73f07123bdbb
18403277b5621add4156448b930801)))),UInt32Impl(0),UInt32Impl(4294967293),UInt16Impl(107),None)),wpkh(586623e8faef5c5d6482ea7d15a895aa18455d2a),UInt64Impl(1523917744202123698),UInt64Impl(12603078646807090937),1 sats/vbyte,BlockTime(UInt32I
mpl(1644019200)),BlockTime(UInt32Impl(1644624000))))
org.sqlite.SQLiteException: [SQLITE_CONSTRAINT_PRIMARYKEY]  A PRIMARY KEY constraint failed (UNIQUE constraint failed: funding_inputs.out_point)
        at org.sqlite.core.DB.newSQLException(DB.java:1030)
        at org.sqlite.core.DB.newSQLException(DB.java:1042)
        at org.sqlite.core.DB.throwex(DB.java:1007)
        at org.sqlite.core.DB.executeBatch(DB.java:832)
        at org.sqlite.core.CorePreparedStatement.executeBatch(CorePreparedStatement.java:64)
        at com.zaxxer.hikari.pool.ProxyStatement.executeBatch(ProxyStatement.java:128)
        at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeBatch(HikariProxyPreparedStatement.java)
        at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$MultiInsertAction.$anonfun$run$18(JdbcActionComponent.scala:542)
        at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement(JdbcBackend.scala:425)
        at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement$(JdbcBackend.scala:420)
        at slick.jdbc.JdbcBackend$BaseSession.withPreparedStatement(JdbcBackend.scala:489)
        at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl.preparedInsert(JdbcActionComponent.scala:511)
        at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$MultiInsertAction.run(JdbcActionComponent.scala:536)
        at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:28)
        at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:25)
        at slick.basic.BasicBackend$DatabaseDef$$anon$3.liftedTree1$1(BasicBackend.scala:276)
        at slick.basic.BasicBackend$DatabaseDef$$anon$3.run(BasicBackend.scala:276)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at java.base/java.lang.Thread.run(Thread.java:831)
```

**Comment by Christewart** [comment]
fixed in #4067
